### PR TITLE
feat: check guest availabilty when rescheduled

### PR DESCRIPTION
**Pull request by Pavan-Chaithanya**
## What does this PR do?

It takes all the attendees and checks their availabilities while rescheduling

/claim https://github.com/calcom/cal.com/issues/16378

- Fixes https://github.com/calcom/cal.com/issues/16378

## Mandatory Tasks (DO NOT REMOVE)

- [X] I have self-reviewed the code (A decent size PR without self-review might be rejected).
- [X] I have added a Docs issue [here](https://github.com/calcom/docs/issues/new) if this PR makes changes that would require a [documentation change](https://docs.cal.com/docs). If N/A, write N/A here and check the checkbox.
- [X] I confirm automated tests are in place that prove my fix is effective or that my feature works.

## How should this be tested?

- Create an event as host and send it to the guest
- Once the guest books a slot in the host calendar, try rescheduling from the host account.
- If events are already on the guest calendar, those slots won't be available to the host while rescheduling.

### CODE REVIEW

File: packages/prisma/schema.prisma [hunk 1]
```diff
@@ -58,8 +58,6 @@ model Host {
   @@index([eventTypeId])
 }
 
-
-
```
Comment (anikdhabal) [inline packages/prisma/schema.prisma:62]: Remove this unnecessary change

Comment (Pavan-Chaithanya) [reply packages/prisma/schema.prisma]: removed

**Comment by BonapartePC** [comment]
@anikdhabal and @Amit91848 Can you review this PR?

**Comment by Pavan-Chaithanya** [comment]
@anikdhabal and @Amit91848 Can you review this PR?

File: packages/trpc/server/routers/viewer/slots/util.ts [hunk 2]
```diff
@@ -423,6 +446,9 @@ export async function getAvailableSlots({ input, ctx }: GetScheduleOptions): Pro
           return usersArray;
         }, [] as (GetAvailabilityUser & { isFixed: boolean })[]);
 
+  if (attendeeUsers.length) {
+    usersWithCredentials.push(...attendeeUsers.map((user) => ({ ...user, isFixed: false })));
```
Comment (Amit91848) [inline packages/trpc/server/routers/viewer/slots/util.ts:450]:
```suggestion
    usersWithCredentials.push(...attendeeUsers.map((user) => ({ ...user, isFixed: true })));
```
Here we want isFixed to be true since want all attendees to strictly be available at the time slot

**Review by Amit91848 — CHANGES_REQUESTED** [review]
Great work @Pavan-Chaithanya. Have left a small suggestion when adding attendeeUsers to users, otherwise really solid changes.
Can you also add another test where user is rescheduling and non of the attendees have cal accounts so that we know for sure this isn't breaking the regular flow 🙏

File: apps/web/test/lib/getSchedule.test.ts [hunk 3]
```diff
@@ -1553,4 +1553,99 @@ describe("getSchedule", () => {
       );
     });
   });
+
+  describe("Rescheduling with multiple attendees", () => {
+    test("correctly identifies unavailable slots when rescheduling with multiple attendees", async () => {
```
Comment (Amit91848) [inline apps/web/test/lib/getSchedule.test.ts:1558]:
```suggestion
    test("correctly identifies unavailable slots when rescheduling with multiple attendees who have Cal accounts", async () => {
```

Comment (Pavan-Chaithanya) [reply apps/web/test/lib/getSchedule.test.ts]: Fixed it

File: packages/trpc/server/routers/viewer/slots/util.ts (hunk 2 above)
Comment (Pavan-Chaithanya) [reply packages/trpc/server/routers/viewer/slots/util.ts]: Made it as true

**Comment by Pavan-Chaithanya** [comment]
> Great work @Pavan-Chaithanya. Have left a small suggestion when adding attendeeUsers to users, otherwise really solid changes. Can you also add another test where user is rescheduling and non of the attendees have cal accounts so that we know for sure this isn't breaking the regular flow 🙏

Added a test where none of the attendees are cal users

**Review by CarinaWolli — CHANGES_REQUESTED** [review]
We should also adjust `handleNewBooking` to only allow booking slots that are returned by `getAvailableSlots` and shown as available

**Comment by Pavan-Chaithanya** [comment]
> We should also adjust `handleNewBooking` to only allow booking slots that are returned by `getAvailableSlots` and shown as available

Made changes to `handleNewBooking` to determine which slots are available

File: packages/features/bookings/lib/handleNewBooking/test/reschedule.test.ts [hunk 4]
```diff
@@ -2193,7 +2531,7 @@ describe("handleNewBooking", () => {
           });
 
           const roundRobinHost2 = getOrganizer({
-            name: "RR Host 2",
```
Comment (Pavan-Chaithanya) [inline packages/features/bookings/lib/handleNewBooking/test/reschedule.test.ts:2196]: I made this change since [this](https://github.com/calcom/cal.com/blob/main/packages/lib/defaultEvents.ts) regex considers this name as three user names due to spaces b/w them.

File: packages/features/bookings/lib/handleNewBooking.ts [hunk 5]
```diff
@@ -475,6 +476,24 @@ async function handler(
     ) {
       throw new HttpError({ statusCode: 403, message: ErrorCode.CancelledBookingsCannotBeRescheduled });
     }
+    const availableSlots = await getAvailableSlots({
```
Comment (Amit91848) [inline packages/features/bookings/lib/handleNewBooking.ts:479]: Not sure if we should be calling getAvailableSlots here. Later on ensureAvailableUsers is also called and both use getUserAvailability internally. Can you check if there is any impact of the performance?

I think you should add guests to the users that we are passing to ensureAvailableUsers.

cc @CarinaWolli would love you thoughts on this as I am not completely sure 🙏

**Review by Amit91848 — CHANGES_REQUESTED** [review]
Left some comments @Pavan-Chaithanya

File: packages/features/bookings/lib/handleNewBooking/test/reschedule.test.ts [hunk 6]
```diff
@@ -690,6 +690,117 @@ describe("handleNewBooking", () => {
...
+                name: booker.name,
+                location: { optionValue: "", value: BookingLocations.CalVideo },
+              },
+            },
+          });
+          const { req } = createMockNextJsRequest({
+            method: "POST",
+            body: mockBookingData,
+          });
+          // Expect await handleNewBooking(req); to throw exception
```
[the remark this answers is not in the document]
Comment (Pavan-Chaithanya) [reply packages/features/bookings/lib/handleNewBooking/test/reschedule.test.ts]: Removed

File: packages/features/bookings/lib/handleNewBooking.ts [hunk 7]
```diff
@@ -475,6 +476,24 @@ async function handler(
...
+      input: {
+        startTime: reqBody.start,
+        endTime: reqBody.end,
+        eventTypeId: eventTypeId,
+        eventTypeSlug: eventTypeSlug,
+        timeZone: reqBody.timeZone,
+        usernameList: dynamicUserList,
```
[the remark this answers is not in the document]
Comment (Pavan-Chaithanya) [reply packages/features/bookings/lib/handleNewBooking.ts]: Removed this logic and moved to `ensureAvailableUsers` function

File: packages/features/bookings/lib/handleNewBooking.ts [hunk 8]
```diff
@@ -475,6 +476,24 @@ async function handler(
...
+        duration: reqEventLength - 1,
```
[the remark this answers is not in the document]
Comment (Pavan-Chaithanya) [reply packages/features/bookings/lib/handleNewBooking.ts]: Removed this logic and moved to `ensureAvailableUsers` function

File: packages/features/bookings/lib/handleNewBooking.ts (hunk 5 above)
Comment (Pavan-Chaithanya) [reply packages/features/bookings/lib/handleNewBooking.ts]: I moved most of the logic to the `ensureAvailableUsers` function since they are similar.

**Comment by Pavan-Chaithanya** [comment]
@anikdhabal @Amit91848 @CarinaWolli Can I get a review for this PR, please?

**Comment by Pavan-Chaithanya** [comment]
@Amit91848 Any update on this?

**Review by CarinaWolli — CHANGES_REQUESTED** [review]
We still want to allow the guests themselves to reschedule outside of their availability. So we only want to check for the guest's availability when the event type owners (or team admins) reschedule. 

See https://github.com/calcom/cal.com/issues/16378#issuecomment-2347342541

File: packages/features/bookings/lib/handleNewBooking/ensureAvailableUsers.ts [hunk 9]
```diff
@@ -64,11 +69,46 @@ export async function ensureAvailableUsers(
...
+            ...u,
+            isFixed: true,
+          };
+        });
+      }
+    }
+  }
+  const totalUsers = [...eventType.users, ...attendeeUsers].filter(
+    (user, index, self) => index === self.findIndex((t) => t.id === user.id)
+  );
```
Comment (Amit91848) [inline packages/features/bookings/lib/handleNewBooking/ensureAvailableUsers.ts:107]: `EnsureAvailableUsers` takes users and checks if those users are present for the timeslot. It is not good practice to modify those users inside the function. This will make it prone to errors in future.

**Review by Amit91848 — CHANGES_REQUESTED** [review]

File: packages/features/bookings/lib/handleNewBooking/ensureAvailableUsers.ts (hunk 9 above)
Comment (Pavan-Chaithanya) [reply packages/features/bookings/lib/handleNewBooking/ensureAvailableUsers.ts]: Made the change

**Comment by Pavan-Chaithanya** [comment]
@Amit91848 @CarinaWolli Addressed all the comments. Can you review the PR?

**Comment by Pavan-Chaithanya** [comment]
Any update on this PR?

File: packages/trpc/server/routers/viewer/slots/util.ts [hunk 10]
```diff
@@ -424,9 +426,44 @@ export async function getAvailableSlots({ input, ctx }: GetScheduleOptions): Pro
...
+      hosts = hosts.filter((host) => host.user.id === originalRescheduledBooking?.userId || 0);
+    }
+
+    const session = ctx?.req ? await getServerSession({ req: ctx?.req }) : null;
+    const userId = session?.user?.id;
+    const isUserEventOwner = isUserReschedulingOwner(userId, originalRescheduledBooking?.userId || 0)
+    const membership = eventType?.team?.members.find((membership) => membership.userId === userId);
+    const isUserTeamAdminOrOwner = membership?.role === MembershipRole.OWNER || membership?.role === MembershipRole.ADMIN;
+    if (isUserEventOwner || isUserTeamAdminOrOwner) {
+      // Get attendees email ids
```
Comment (Amit91848) [inline packages/trpc/server/routers/viewer/slots/util.ts:449]: NIT: understandable by code no need to add comment

```suggestion
```

File: packages/trpc/server/routers/viewer/slots/util.ts [hunk 11]
```diff
@@ -424,9 +426,44 @@ export async function getAvailableSlots({ input, ctx }: GetScheduleOptions): Pro
...
+      eventType.schedulingType === SchedulingType.ROUND_ROBIN
+    ) {
+      hosts = hosts.filter((host) => host.user.id === originalRescheduledBooking?.userId || 0);
+    }
+
+    const session = ctx?.req ? await getServerSession({ req: ctx?.req }) : null;
+    const userId = session?.user?.id;
+    const isUserEventOwner = isUserReschedulingOwner(userId, originalRescheduledBooking?.userId || 0)
+    const membership = eventType?.team?.members.find((membership) => membership.userId === userId);
+    const isUserTeamAdminOrOwner = membership?.role === MembershipRole.OWNER || membership?.role === MembershipRole.ADMIN;
```
Comment (Amit91848) [inline packages/trpc/server/routers/viewer/slots/util.ts:447]: This covers the scenario for team owners but would fail if current user is org owners but they are not in the same team as event owner.

**Review by Amit91848 — CHANGES_REQUESTED** [review]
Great work so far left some suggestions 🙏

File: packages/trpc/server/routers/viewer/slots/util.ts (hunk 10 above)
Comment (Pavan-Chaithanya) [reply packages/trpc/server/routers/viewer/slots/util.ts]: Removed

**Comment by Pavan-Chaithanya** [comment]
@Amit91848 Thanks for the reviews so far. Any update on this PR?

**Comment by Pavan-Chaithanya** [comment]
@Amit91848 Gentle reminder

**Comment by anikdhabal** [comment]
Hey @Pavan-Chaithanya don't panic, we will review it shortly

**Comment by anikdhabal** [comment]
Hey @Pavan-Chaithanya, I appreciate your work and patience, but since this is a bounty issue, we need to choose one. So going with this https://github.com/calcom/cal.com/pull/16416
